Create wrapper and logger for Serial com

m3d opened the pull request:
This is the first step to creation of `LogSerial` described in in https://github.com/robotika/osgar/pull/44#issuecomment-364014286

I am not sure if to modify `drivers\gps.py` as the first example or in separate PR? The same holds for replay and handling `process` and communication queues.

m3d commented:
I added experiment with modified GPS too. We can split it if necessary. It is rather "sketch" how it could be integrated.

Issues:
- the `output` should be just one function with one parameter? On `LogSerial` side I call `self.output.put`, but I am inclined to change it to just call `self.output` as in `GPS`
- the content of packet could be (timestamp, stream_id, data), so the receiver can compare how old the data are (or rather for post-analysis), switch parser based on stream_id, and use data. For logger the data have to be already serialized (bytes), but do we want them serialized in inter-module "communication" (via Queue)?
- `process` is meant to do the work, to generate output packet or nothing (None) yet. I am thinking to cut off all `Thread` code and use it externally (Mixin?). Or do you prefer to have the classes inherited from Thread and call only internal methods? I suppose that `run()` will be for all modules the same
- I would keep `timeout` for reading from Queue, but send it to receiver. This means that receiver can receive data (say coordinate) or None (timeout on read). One motivation is to be able to shutdown the whole system correctly and also to log suspicious delays. OK?

There is still issue with binary/text data, which will probably follow.

m3d commented:
One more commit for experiment with assert processing. This code should be general for all (?) modules but I am not sure where it should live??

m3d commented:
There is one more proposal how could binary GPS messages look like when numpy structured arrays are used (https://docs.scipy.org/doc/numpy/user/basics.rec.html)

zwn commented:
I'd take yet another path.

I could create Bus and everything would go through it. The Bus would also log itself and internally would do all necessary locking. Its usage would look something like this:

In serial driver:

```python
while self.should_run.isSet():
    data = self.com.read(1024)
    self.bus.publish(self.stream_id, data)
```

In gps driver (well it would no longer be a driver but a generic module communicating only through the bus):

```python
while self.should_run.isSet():
    dt, id, data = self.bus.listen(self.stream_id_in)
    out_data = self.process(data)
    self.bus.publish(self.stream_id_out, out_data)
```

As for the low level implementation
 * I am not very fond of inheritance in python. The less the better. The more Thread is separated from the actual driver/module, the better.
 * I would send messages "as is", no pre-serialization. It the Bus wants to log, just use pickle for now.
 * If something can be called as a function, I'd rather do that instead of calling it as a method (output vs output.put)

zwn commented:
The key to my suggested approach is that the logging facility and the communication facility is completely outside of the driver/module code. Also note that the logging is irreversibly connected with the communication. If we decide one of the modules should live in a separate process it should be easy to do "from the outside".

zwn commented:
To allow for a proper shutdown I would investigate the possibility to throw exceptions from the Bus methods (listen, publish) instead of using Event object in a Thread. That would further separate the modules from the Thread interface allowing us (in the future) to run modules using Multiprocessing instead of Thread.

m3d commented:
OK, I added sketch of `Bus` (almost dummy). Where I stopped was in `listen(stream_id)` ... I guess that this should be blocking until you receive data or `BusShutdown` exception, correct? What should be happening with published message if you did not call `listen` (during time between two `listen` calls)? I.e. is there dedicated queue with messages just for you? If yes, then there is need to register and call listen with that registration ID??

If nobody is interested in particular `publish` message, it will be just logged and dropped, right?

As we discussed over email, I would support `data` as bytes, numpy structured arrays with dtype, and fall back would be now `bytes(str())` with `literal_eval` for decoding.

zwn commented:
Ok, let me iterate on my proposal.

Each module should get its own local handle to The Bus. This handle would translate channel ids and would allow calling `bus.listen()` which would return whatever message arrives first on one of its input channels. A generic client would work like this:

```python
try:
  while(True):
    id, timestamp, data = bus.listen()
    if id == 0:
      bus.publish(0, output_data)
    ...
except BusShutdown:
  pass
```

The channel ids are local to the client. `bus.publish` outputs message on (local) channel 0 and finds out who is interested (from config) and enqueues this message to all receivers (which wakes them up from their bus.listen). `bus.publish` does not block, only `bus.listen` blocks.
 
So for example Spider has two outputs - odometry (0) and stop button (1). Then we might have Localize with two inputs - gps (0) and odometry(1). Then Gps with one output (0). And finally Shutdown with one input (0). In the config we say that Spider:0 (or Spider:odometry) gets sent to Gps:1 (or Gps:odometry) etc.

The implementation might involve using some BusHandles that share the logger.

```python
class BusHandle:
  def publish(self, channel, data):
    with self.lock:
      timestamp = self.log.append((self.log[channel], data))
      for queue, channel in self.out[channel]:
        queue.put((channel, timestamp, data))
  def listen(self):
    channel, timestamp, data = self.queue.get()
    return channel, timestamp, data
```

The bus handles would be given to the clients "from outside" already properly preconfigured according to config info.

Does it make sense? It should be flashed out more during the implementation (maybe channels should use names for better readability).

m3d commented:
First of all thanks for detail description + extra explanation over telco last night. Here is my first approximation now, without internal mapping yet (I plan to do it, just the time was too limited today :(

So far there were two details for discussion:
- should we really pass the whole `queue` to other handlers or just methods `put` and `get`? I mean that other handlers should not look at others queues except add new data?
- the order of triplets - I used (timestamp, stream_id, data) and you use (channel, timestamp, data). Mine view was defined by order in log file (ordered by time), while yours is probably based on switch view after `listen()` necessary to decide what source of data is it (?). I would keep the same order in both situations to avoid mismatches.

zwn commented on drivers/bus.py:
```diff
@@ -3,17 +3,23 @@
-            pass
+            stream_id = channel  # TODO local maping of indexes
+            timestamp = self.logger.write(stream_id, data)
+            if channel in self.out:
```
It the code tries to publish on undefined/unallocated channel, it should fail IMO.

zwn commented on drivers/bus.py:
```diff
@@ -3,17 +3,23 @@
+class BusHandler:
+    def __init__(self, logger, input_queue=None, out={}):
         self.logger = logger
+        self.queue = input_queue
```
I'd create the queue just here. There is going to be one per Handler anyway.

zwn commented on drivers/bus.py:
```diff
@@ -3,17 +3,23 @@
-class Bus:
-    def __init__(self, logger):
+class BusHandler:
+    def __init__(self, logger, input_queue=None, out={}):
```
I think we need input channels here as well. But that decision depends on a code that has not yet been written 😉

zwn commented on drivers/test_bus.py:
```diff
@@ -1,22 +1,37 @@
-        bus = Bus(logger)
+        bus = BusHandler(logger)
         bus.publish(1, b'some binary data')
         logger.write.assert_called_once_with(1, b'some binary data')
```
I think here it should catch exception instead since no outputs are defined. In order for this not to fail it should contain this:

```python
bus = BusHandler(logger, out={1:[]})
```

That defines output channel named `1` that is not attached anywhere.

zwn commented:
> First of all thanks for detail description + extra explanation over telco last night. Here is my first approximation now, without internal mapping yet (I plan to do it, just the time was too limited today :(

You are welcome 😄 

> should we really pass the whole queue to other handlers or just methods put and get? I mean that other handlers should not look at others queues except add new data?

I don't have a strong opinion. You can get to the queue if you really want to just from the method anyway so I would not consider that to be the criteria to decide. I'd would do what ever makes the code more readable (I am not sure what that is yet). Having the queue there seemed sensible at the time since the BusHandler is the one handling the queues anyway. 

> the order of triplets - I used (timestamp, stream_id, data) and you use (channel, timestamp, data). Mine view was defined by order in log file (ordered by time), while yours is probably based on switch view after listen() necessary to decide what source of data is it (?). I would keep the same order in both situations to avoid mismatches.

That was not a conscious decision on my side. Just keep what you are using. Ordering by timestamp is a good argument 👍

zwn commented:
As for config, here some quick brainstorming:

```json
{
  "can": {
    "driver": "serial",
    "out": ["raw"],
    "init": ["/dev/tty01", "..."]
  },
  "spider": {
    "driver": "spider",
    "in": ["can.raw"],
    "out" : ["odometry", "stop"],
    "init": []
  }
}
```

I think it might even be self-describing. If it is not, let me know.

zwn commented on drivers/bus.py:
```diff
@@ -0,0 +1,28 @@
+"""
+  Interna bus for communitaion among modules
```
Spelling.

m3d commented:
There is my first attempt to integrate handler in GPS and LogSerial (so far in "run" mode only). As usual there are couple details [OT I am aware that you want to remove `Thread` part completely from modules, this will be probably next step].

I am still not sure about "timeout" in `listen` method. Somehow I think it should be available there. If you are LogSerial, you do not know anything about receiving data, but if you are GPS you know, that you should get new position at least every second (maybe not valid, but the data should come). Similarly if you parse data from Spider ECU - there should be update every 50ms. And if it is not maybe you should react on that timeout and log this problem, that in given real time you still did not get expected message??

The second issue is termination. You wanted to raise exception ... but the module is probably blocked in `Queue.get()`? I think that in https://github.com/robotika/katarina (but I do not see it there now) termination of external processing by simply passing `None` into queue. I.e. instead of expected triples (timestamp, channel, data) you get `None`?? OK/NOK? BTW should this be logged?

Finally I hacked `LogWriterEx` to extend existing `LogWriter` and handle also numpy arrays. Where to put this class? Somehow I would like to keep the logger.py simple, with just raw bytes, so this could be serialize wrapper ... and if I remember correctly you do not want it in `BusHandler`, right?

zwn commented on drivers/gps.py:
```diff
@@ -37,19 +37,16 @@ def str2ms(s):
         self.should_run.set()
 
-        self.logger = logger
+        self.handler = handler
```
Can we call this `bus` instead of `handler`?

zwn commented on drivers/gps.py:
```diff
@@ -104,21 +99,34 @@ def request_stop(self):
         time.sleep(2)
         device.request_stop()
         device0.request_stop()
+        device0.handler.publish(1, b'')  # hack - make sure the GPS blocked read is unblocked
```
Can we start using names instead of numbers?

zwn commented:
> I am still not sure about "timeout" in listen method. Somehow I think it should be available there.

This discussion is getting really long and the changes are quite difficult to review already and getting harder the more commits we have. I'd rather do less in this PR than more. I agree there seems to be some utility to having timeouts but it would further complicate log replay. I'd rather findd something that can be merged asap and continue in different PRs.

> The second issue is termination. You wanted to raise exception ... but the module is probably blocked in Queue.get()? I think that in https://github.com/robotika/katarina (but I do not see it there now) termination of external processing by simply passing None into queue. I.e. instead of expected triples (timestamp, channel, data) you get None?? OK/NOK? BTW should this be logged?

If you use None you will not be able to keep `timestamp, channel, data = self.queue.get()` syntax. But I agree we need some kind of cookie, yes. I don't see the need to log it. When that need arises it can easily be added.

> Finally I hacked LogWriterEx to extend existing LogWriter and handle also numpy arrays. Where to put this class? Somehow I would like to keep the logger.py simple, with just raw bytes, so this could be serialize wrapper ... and if I remember correctly you do not want it in BusHandler, right?

Right, serialization does not belong to communication class. I'd just create a function that returns bytes representation of its argument and called it in LogWriter. I would not create a subclass of LogWriter for this.

Anyway, the less we have in this PR the better. The task now should be to find something compact that can be merged now and continue in different PRs (timeouts, numpy, replay, loading from config file).

zwn commented:
We do not need to merge it to master. Just the PRs need to be simpler/smaller/shorter. Thanks.

m3d commented:
Squashed and rebased. There are still several issues TODO but maybe let's merge it and then refactor it in new PR. OK?
